src/wsproxy.ml

```
@@ -15,124 +15,115 @@
-         let buffer = String.make 16384 '\000' in
-         Lwt.catch
-           (fun () ->
-              let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
-              Lwt_unix.recv_msg ~socket:fd_sock2 ~io_vectors:[iov])
-           (fun e ->
-              Printf.printf "Caught exception: %s\n" (Printexc.to_string e);
-              ignore_exn (fun () -> Lwt_unix.close fd_sock2) () >>= fun () ->
-              Lwt.return (0,[]))
-         >>= fun (len,newfds) ->
-         Lwt_unix.close fd_sock2 >>= fun () ->
-         Lwt.catch
-           (fun () ->
-              let fdh :: fdt = newfds in
-              let msg = String.sub buffer 0 len in
-              let fdno = Printf.sprintf "%d" (Obj.magic fdh) in
-              List.iter (fun fd -> Printf.printf "closing fds: %d\n%!" (Obj.magic fd)) fdt;
-              List.iter (fun fd -> try Unix.close fd with _ -> ()) fdt;
-              Printf.printf "About to fixup: %s\n%!" fdno;
-              let fd = Lwt_unix.of_unix_file_descr fdh in
-              Lwt_unix.setsockopt fd Lwt_unix.SO_KEEPALIVE true;
-              handler fd msg)
-           (fun e ->
-              Printf.printf "Caught exception: %s\n" (Printexc.to_string e);
-              List.iter (fun fd -> Printf.printf "closing fd: %d\n%!" (Obj.magic fd)) newfds;
-              List.iter (fun fd -> try Unix.close fd with _ -> ()) newfds;
-              Lwt.return ()) 
-         >>= fun _ ->
-         loop ()
-      )
+         Lwt_unix.accept fd_sock
+         >>= fun (fd_sock',_) ->
+         (* Background thread per connection *)
+         let _ =
+           let buffer = String.make 16384 '\000' in
+           with_fd fd_sock'
+             (fun fd ->
+                let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
+                Lwt_unix.recv_msg ~socket:fd ~io_vectors:[iov])
+           >>= fun (len, newfds) ->
+           match newfds with
+           | [] -> Lwt.return_unit
```

Should we log something here or throw an exception, if this is an unexpected situation?